website/docs/externaldata.md

```diff
@@ -0,0 +1,154 @@
+---
+id: externaldata
+title: External Data
+---
+
+> This feature is still in alpha stage, so the final form can still change (feedback is welcome!). Mutation is not yet supported with external data.
+
+## Motivation
+
+Gatekeeper provides various means to mutate and validate Kubernetes resources. However, in many of these scenarios this data is either built-in, static or user-defined. With external data feature, we are enabling Gatekeeper to interface with various external data sources, such as image registries, using a provider-based model.
+
+A similar way to connect with an external data source can be done today using OPA's built-in `http.send` functionality. However, there are limitations to this approach.
+- Gatekeeper does not support rego policies for mutation, which can not use the OPA http.send built-in function
+- There are unknown factors when the proxy endpoint is not responsive, doesn't exist/down or fails. These are addressed by external data interfaces such as timeout.
```

Comment:
Also, potential security risks... if you don't trust the author of a constraint template, then `http.Send()` is a bad idea, as it gives them network access.

If you believe the CT author is safe, they still need to carefully consider how their Rego is written to avoid injection attacks. External Data helps with this by:

* Restricting which hosts a user can access
* Providing an interface for making requests, which allows us to better handle things like escaping strings.

Also, potential for performance benefits, as we can more directly control caching. In the case of mutation we can batch requests via lazy evaluation. For validation we make batching easier via function design.

Finally, using external data primitives makes it easier to keep track of data sources by providing a common format for registering and accessing them.

Those are the other benefits that come to mind. I may be missing some.

I'd avoid listing failure handling/timeouts/etc. as a way `http.Send()` doesn't fit, since it could probably be extended to provide these. The biggest differentiator is that `http.Send()` will send any request to any host, whereas external data wont (along with other restrictions on how the request is formatted/made). This makes it less flexible, but also provides us space to offer value as a platform, like mitigating some of the above security concerns.

Reply:
Thank you! updated